pkg/odo/cli/storage/create.go

```diff
@@ -30,121 +25,69 @@ var (
+		Cmd:              cmd,
+		DevfilePath:      component.DevfilePath,
+		ComponentContext: o.componentContext,
+	})
 
-		if o.storageSize == "" || o.storagePath == "" {
-			return fmt.Errorf("\"size\" and \"path\" flags are required for s2i components")
-		}
+	if err != nil {
+		return err
 	}
 
 	if len(args) != 0 {
 		o.storageName = args[0]
 	} else {
-		o.storageName = fmt.Sprintf("%s-%s", o.componentName, util.GenerateRandomString(4))
+		o.storageName = fmt.Sprintf("%s-%s", o.Context.LocalConfigProvider.GetName(), util.GenerateRandomString(4))
 	}
 
-	o.applyDevfileStorageDefault()
+	o.storage = localConfigProvider.LocalStorage{
+		Name: o.storageName,
+		Size: o.storageSize,
+		Path: o.storagePath,
+	}
 
-	return
-}
+	o.Context.LocalConfigProvider.CompleteStorage(&o.storage)
 
-func (o *StorageCreateOptions) applyDevfileStorageDefault() {
-	if o.storageSize == "" {
-		o.storageSize = adapterCommon.DefaultVolumeSize
-	}
-	if o.storagePath == "" {
-		// acc to the devfile schema, if the mount path is absent; it will be mounted at the dir with the mount name
-		o.storagePath = "/" + o.storageName
-	}
+	return
 }
 
-// Validate validates the StorageCreateOptions based on completed values
-func (o *StorageCreateOptions) Validate() (err error) {
-	if o.isDevfile {
-		return
-	}
-	// validate storage path
-	return o.LocalConfigInfo.ValidateStorage(o.storageName, o.storagePath)
+// Validate validates the CreateOptions based on completed values
+func (o *CreateOptions) Validate() (err error) {
+	// validate the storage
+	return o.LocalConfigInfo.ValidateStorage(o.storage)
 }
 
-func (o *StorageCreateOptions) devfileRun() error {
-	devFile, err := devfile.ParseAndValidate(o.devfilePath)
-	if err != nil {
-		return err
-	}
-	err = validate.ValidateDevfileData(devFile.Data)
-	if err != nil {
-		return err
-	}
-
-	err = devFile.Data.AddVolume(devfilev1.Component{
-		Name: o.storageName,
-		ComponentUnion: devfilev1.ComponentUnion{
-			Volume: &devfilev1.VolumeComponent{
-				Volume: devfilev1.Volume{
-					Size: o.storageSize,
-				},
-			},
-		},
-	}, o.storagePath)
-
-	if err != nil {
-		return err
-	}
-	err = devFile.WriteYamlDevfile()
+// Run contains the logic for the odo storage create command
+func (o *CreateOptions) Run() (err error) {
+	err = o.Context.LocalConfigProvider.CreateStorage(o.storage)
 	if err != nil {
 		return err
 	}
-	return nil
-}
-
-// Run contains the logic for the odo storage create command
-func (o *StorageCreateOptions) Run() (err error) {
-	if o.isDevfile {
-		err := o.devfileRun()
-		if err != nil {
-			return err
-		}
-	} else {
-		_, err := o.LocalConfigInfo.StorageCreate(o.storageName, o.storageSize, o.storagePath)
-		if err != nil {
-			return err
-		}
-	}
 
-	storageResultMachineReadable := storage.GetMachineReadableFormat(o.storageName, o.storageSize, o.storagePath)
+	storageResultMachineReadable := storage.GetMachineReadableFormat(o.storage.Name, o.storage.Size, o.storage.Path)
```

Comment:
I understand that this is not something that's been touched by your PR but would it make sense to change the signature of `GetMachineReadableFormat` to just take the storage's `CreateOptions` instead of taking three `string`s?

I'm asking this because if we modify this JSON output in future to add/remove something from it, we will have to change the function signature. And that would break unit tests. Instead, if we modify it to just use one parameter, the storage `CreateOptions`, unit tests can help us point when something's not working as expected.

Reply:
Also, are we evaluating this if it's needed only in the case when the user has request for JSON output? Shouldn't this be in the block of `if log.IsJSON()` instead?

Reply:
> I understand that this is not something that's been touched by your PR but would it make sense to change the signature of GetMachineReadableFormat to just take the storage's CreateOptions instead of taking three strings?

That would create a import cycle.

> Also, are we evaluating this if it's needed only in the case when the user has request for JSON output? Shouldn't this be in the block of if log.IsJSON() instead?

Fixed